metadata/src/utils/retain.rs

```diff
@@ -8,45 +8,185 @@ use crate::{
+            extrinsic.call_ty,
+            extrinsic.signature_ty,
+            extrinsic.extra_ty,
+        ]);
+
+        for signed in &extrinsic.signed_extensions {
+            ids.push(signed.extra_ty);
+            ids.push(signed.additional_ty);
+        }
+        for id in ids {
+            self.seen_ids.insert(id);
+        }
     }
 
-    for constant in pallet.constants.values() {
-        type_ids.insert(constant.ty);
+    /// Collect all type IDs needed to represent the runtime APIs.
+    fn collect_runtime_api_types(&mut self, metadata: &Metadata, api: &RuntimeApiMetadataInner) {
+        for method in api.methods.values() {
+            for input in &method.inputs {
+                self.insert_and_collect(metadata, input.ty);
+            }
+            self.insert_and_collect(metadata, method.output_ty);
+        }
+    }
+
+    /// Collect all type IDs needed to represent the provided pallet.
+    fn collect_pallet_types(&mut self, pallet: &PalletMetadataInner, metadata: &Metadata) {
+        let mut type_ids = Vec::new();
+        if let Some(storage) = &pallet.storage {
+            for entry in storage.entries() {
+                match entry.entry_type {
+                    StorageEntryType::Plain(ty) => {
+                        type_ids.push(ty);
+                    }
+                    StorageEntryType::Map {
+                        key_ty, value_ty, ..
+                    } => {
+                        type_ids.push(key_ty);
+                        type_ids.push(value_ty);
+                    }
+                }
+            }
+        }
+
+        if let Some(ty) = pallet.call_ty {
+            type_ids.push(ty);
+        }
+
+        if let Some(ty) = pallet.event_ty {
+            type_ids.push(ty);
+        }
+
+        for constant in pallet.constants.values() {
+            type_ids.push(constant.ty);
+        }
+
+        if let Some(ty) = pallet.error_ty {
+            type_ids.push(ty);
+        }
+        for id in type_ids {
+            self.insert_and_collect(metadata, id);
+        }
     }
 
-    if let Some(ty) = pallet.error_ty {
-        type_ids.insert(ty);
+    // Collect types referenced inside outer enum
+    fn collect_variants_in_type<F>(&mut self, metadata: &mut Metadata, id: u32, mut name_filter: F)
+    where
+        F: FnMut(&str) -> bool,
+    {
+        let ty = {
+            metadata
+                .types
+                .types
+                .get_mut(id as usize)
+                .expect("Metadata should contain enum type in registry")
+        };
+
+        // Redo the thing above but keep filtered out variants if they reference types that we intend to keep
+        let TypeDef::Variant(variant) = &mut ty.ty.type_def else {
+            panic!("Metadata type is expected to be a variant type");
+        };
+
+        // If the type was not referenced earlier we can safely strip some of the variants
+        if self.seen_ids.insert(id) {
```

Comment:
```suggestion
        if self.seen_ids.contains(id) {
```

Inserting makes the result of calling `collect_variants_in_type` depend on the order in which you call it. So I wonder whether there is a need to insert here?

Reply:
I'm not sure i'm following, we still need to add the enum type to the set and we don't do a deep traversal for outerenums anyways so insertion order shouldn't matter

Reply:
Maybe this is one of the things I find harder to follow; I guess I want to separate the funciton that "determines whether to keep all of the variants" from functions that mutate and add type IDs to some set.